vid/src/advz.rs

```diff
@@ -809,65 +826,107 @@ where
-        let pre_fft_len = coeffs_vec.len();
-        EvaluationDomain::ifft_in_place(domain_ref, &mut coeffs_vec);
-
-        // sanity check: the fft did not resize coeffs.
-        // If pre_fft_len != self.recovery_threshold * self.multiplicity
-        // then we were not given the correct number of coeffs. In that case
-        // coeffs.len() could be anything, so there's nothing to sanity check.
-        if pre_fft_len == chunk_size {
-            assert_eq!(coeffs_vec.len(), pre_fft_len);
+        let mut evals_vec: Vec<_> = evals.map(|c| *c.borrow()).collect();
+        let pre_fft_len = evals_vec.len();
+        if pre_fft_len > domain_size {
+            return Err(VidError::Internal(anyhow::anyhow!(
+                "number of evals {} exceeds domain_size {}",
+                pre_fft_len,
+                domain_size
+            )));
+        }
+        let domain = Self::eval_domain(domain_size)?;
+
+        domain.ifft_in_place(&mut evals_vec);
+
+        // sanity: the fft did not resize evals. If pre_fft_len < domain_size
+        // then we were given too few evals, in which case there's nothing to
+        // sanity check.
+        if pre_fft_len == domain_size && pre_fft_len != evals_vec.len() {
+            return Err(VidError::Internal(anyhow::anyhow!(
+                "unexpected output resize from {pre_fft_len} to {}",
+                evals_vec.len()
+            )));
         }
 
-        DenseUVPolynomial::from_coefficients_vec(coeffs_vec)
+        Ok(DenseUVPolynomial::from_coefficients_vec(evals_vec))
     }
 
-    fn polynomial<I>(&self, coeffs: I) -> KzgPolynomial<E>
-    where
-        I: Iterator,
-        I::Item: Borrow<KzgEval<E>>,
-    {
-        Self::polynomial_internal(
-            &self.eval_domain,
-            (self.recovery_threshold * self.multiplicity) as usize,
-            coeffs,
-        )
+    fn min_multiplicity(&self, payload_byte_len: usize) -> VidResult<u32> {
+        let elem_bytes_len = bytes_to_field::elem_byte_capacity::<<E as Pairing>::ScalarField>();
+        let elems: u32 = payload_byte_len
+            .div_ceil(elem_bytes_len)
+            .try_into()
+            .map_err(vid)?;
+        if self.recovery_threshold * self.max_multiplicity < elems {
+            // payload is large. no change in multiplicity needed.
+            return Ok(self.max_multiplicity);
+        }
+
+        // payload is small: choose the smallest `m` such that `0 < m <
+        // multiplicity` and the entire payload fits into `m *
+        // recovery_threshold` elements.
+        let m = elems.div_ceil(self.recovery_threshold).max(1);
+
+        // TODO TEMPORARY: enforce power-of-2
+        // https://github.com/EspressoSystems/jellyfish/issues/668
+        //
+        // Round up to the nearest power of 2.
+        //
+        // After the above issue is fixed: delete the following code and return
+        // `m` from above.
+        if m <= 1 {
```

Comment:
> nit: guess we can either drop the `max(1)` or use `m==1` here.

We could but I won't. I don't want to drop the `max(1)` because in the future the subsequent `if` statement will be deleted and I don't want that to break this method. OTOH there's no harm in leaving it `m<=1`. In fact, that's more robust to changes elsewhere; I always know that if `m` is 0 or 1 then we should return 1.

> Follow-up thought (not to be addressed here): In general how do we deal with empty payloads/blocks aka. `payload_byte_len==0` in VID? Is it allowed (for downstream components) to call disperse with empty payload? If this is the case, then should we make an effort to "return early" by caching the result of the empty payload disperse and avoid doing unnecessary work?

We do allow zero-length payloads, and tests check for that. (That's how I came to put `max(1)` above.) In an ideal world we would handle the special case of zero-length payload separately, but I don't want to take the time for that.